arch: aarch32: linker command file fixes: .bss/.noinit sections, z_mapped_start marker

ibirnbaum opened the pull request:
This is a follow up to #53262, which still lacked the adjustment of the .noinit section's position within the binary by the time the PR went stale.

Adjust the linker command file so that the .bss and .noinit sections are placed at the end of the resulting binary. Until now, those sections have been located somewhere in the middle of the binary, so that the inclusion of structures like statically defined heaps or large zero- initialized arrays reflected 1:1 in the resulting binary's size. Even for a stripped binary, such data was included in full as the linker couldn't omit it due to subsequent sections within the binary.

This fix has been tested with a 32 MB statically allocated heap and a 32 MB uint8 zero-initialized array. Both structures are clearly identifiable in the memory consumption statistics, however, the final binary's size is unaffected by their inclusion.

ibirnbaum committed:
arch: aarch32: place .bss, .noinit sections at the end of the binary

This is a follow up to #53262, which still lacked the adjustment of the
.noinit section's position within the binary by the time the PR went
stale.

Adjust the linker command file so that the .bss and .noinit sections
are placed at the end of the resulting binary. Until now, those sections
have been located somewhere in the middle of the binary, so that the
inclusion of structures like statically defined heaps or large zero-
initialized arrays reflected 1:1 in the resulting binary's size. Even
for a stripped binary, such data was included in full as the linker
couldn't omit it due to subsequent sections within the binary.

This fix has been tested with a 32 MB statically allocated heap and
a 32 MB uint8 zero-initialized array. Both structures are clearly
identifyable in the memory consumption statistics, however, the final
binary's size is unaffected by their inclusion.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

ibirnbaum commented:
cross-reference to #53262

povergoing commented on include/zephyr/arch/arm/aarch32/cortex_a_r/scripts/linker.ld:
```diff
@@ -329,6 +301,33 @@ SECTIONS
+         */
+        . = ALIGN(4);
+        __bss_start = .;
+        __kernel_ram_start = .;
```
In this way, should `__kernel_ram_start` be put above the `.data` section? Otherwise, `__kernel_ram_start` would not include the data section, right?

ibirnbaum replied:
Let me check where this marker is used, you likely have a point there. This was missed by the author of #53262 as well.

Also, I noticed last night that with regards to the linker command file, the issue described in #51024 also still persists. We should preferrably fix this as well before going for the A/R/M separation.

carlocaione commented on include/zephyr/arch/arm/aarch32/cortex_a_r/scripts/linker.ld:
```diff
@@ -266,34 +266,6 @@ SECTIONS
-         */
-        . = ALIGN(4);
-        __bss_start = .;
-        __kernel_ram_start = .;
```
I think the region between `__kernel_ram_start` and `__kernel_ram_end` should also include the BSS so that is included in the MPU region mapping.

ibirnbaum replied:
done.

carlocaione commented on include/zephyr/arch/arm/aarch32/cortex_a_r/scripts/linker.ld:
```diff
@@ -266,36 +266,9 @@ SECTIONS
-
     SECTION_DATA_PROLOGUE(_DATA_SECTION_NAME,,)
     {
+        __kernel_ram_start = .;
```
can you move this outside the data section?

ibirnbaum replied:
I can test that, sure. I've had a case before though where putting a marker outside a section using the "." operator produced the correct value for the marker, however, the following section's start didn't fit the marker due to alignment. The trouble was that the following section aligned itself to a location *before* the marker outside of the section. That was... weird. I'll see if I still get the same values for the markers if the start marker is outside of the section.

ibirnbaum replied:
Done. I looked it up, the issue I was referring to in my last comment was #41513, btw. In that case, the issue was having a '.'-based marker outside of a region, not outside of a section. So moving `__kernel_ram_start` outside of the data section produces valid results.

ibirnbaum commented:
I've moved the __kernel_ram_start marker out of the bss section and placed it within the first section that originally came right after the bss section, which is the data section. Therefore, data/bss/noinit are now all located between the __kernel_ram_start and __kernel_ram_end markers, but with the adjusted section order. Two test builds of the "philosophers" sample with both revisions of the linker command file produce identical values for the start/end/size markers, as expected, they just differ regarding the order of the sections.

One thing I noticed along the away: I haven't checked who introduced this change and when, but there's now an #include <zephyr/linker/ram-end.ld> statement towards the end of the linker command file which sets _image_ram_end/z_mapped_end. This happens *after* the OCM declaration. In probably 99% of all cases, the OCM is configured to be located at address 0x0, as this matches the default configuration of the exception vectors' location. However, if I configure the OCM to be located at the high address by changing the corresponding 'chosen' entry, the OCM is located at 0xFFFC0000, beyond all RAM, all AXI memory space, all integrated peripherals and the system control registers. Although the OCM is in a separate group, I'd better test if using the high OCM location somehow messes up the _image_ram_end/z_mapped_end markers.

povergoing commented on include/zephyr/arch/arm/aarch32/cortex_a_r/scripts/linker.ld:
```diff
@@ -266,36 +266,9 @@ SECTIONS
-
     SECTION_DATA_PROLOGUE(_DATA_SECTION_NAME,,)
     {
+        __kernel_ram_start = .;
```
The original `__kernel_ram_start = .;` under `.bss` which aligns to `BSS_ALIGN`. This is for the MPU or MMU region settings (MPU requires the address to be aligned to 32-byte or 64-byte, MMU requires 4k). I think you should also add a `_ALIGN` into `SECTION_DATA_PROLOGUE(_DATA_SECTION_NAME,,)`, since data section is the first one. Maybe rename the BSS_ALIGN to DATA_ALIGN or a more generic name, KERNEL_RAM_ALIGN?

povergoing replied:
Anyway, __kernel_ram_start should align to `_region_min_align`

ibirnbaum commented:
I've replaced BSS_ALIGN with a more generic SECTION_ALIGN.

With this revision of the linker command file, I get the follwing memory map when building the philosophers sample for qemu_cortex_a9:

![memmap](https://github.com/zephyrproject-rtos/zephyr/assets/61015245/64084c22-0e0d-48ac-9063-e65f6e802be2)

This looks plausible to me & matches alignment requirements, at least for a MMU-based system.
As mentioned before, z_mapped_start is still in the wrong place (should be 0010 0000), that's the subject of the afforementioned PR #51024. Should I add this fix as a 2nd commit, or should this be a separate PR?
Is it intentional that the markers __data_load_start and __data_region_load_start have no corresponding end markers?

microbuilder commented:
> > Can you add that fix here as well?
> 
> Within the same commit? Sure, but in that case, the title of the PR wouldn't be 100% accurate any more, as we're then mixing the issue of section placement in order to reduce binary size with a misplaced marker indicating to the memory subsystem which memory is available for mapping via the MMU.
> 
> If that's not an issue, I'll go ahead and include the fix, referencing the original PR.

Adding a second commit and making the PR title more generic seems fine to me, to make both changes here.

microbuilder reviewed (changes requested):
This isn't a breaking change, but it feels like it should be in the release notes for Aarch32. Do you mind adding a commit with one or two lines in the 3.5 release notes, highlighting the change and the reason for it?

povergoing commented:
> he syntax error comes from the `ALIGN_WITH_INPUT ALIGN(_region_min_align)` part, if the section alignment specification is removed from linker.ld, `bss (NOLOAD) : ALIGN_WITH_INPUT {...}` seems to by syntactically correct. However, with XIP disabled, the generated output `bss (NOLOAD) : ALIGN(_region_min_align)` is valid syntax as well. Could it be that `ALIGN_WITH_INPUT` and `ALIGN(_region_min_align)` are mutually exclusive and somehow XIP and non-XIP aren't properly separated here?

Does that mean the original code also has this issue?

ibirnbaum commented:
> Does that mean the original code also has this issue?

Yes. Take the unmodified linker command file from the 3.4 code base, build some sample for qemu_cortex_a9, run menuconfig, enable XIP, re-build and you'll get the syntax error, caused by the statement `bss (NOLOAD) : ALIGN_WITH_INPUT ALIGN(_region_min_align)`, with the linker pointing to the generated linker_zephyr_pre0.cmd, line 207.

povergoing commented:
> > Does that mean the original code also has this issue?
> 
> Yes. Take the unmodified linker command file from the 3.4 code base, build some sample for qemu_cortex_a9, run menuconfig, enable XIP, re-build and you'll get the syntax error, caused by the statement `bss (NOLOAD) : ALIGN_WITH_INPUT ALIGN(_region_min_align)`, with the linker pointing to the generated linker_zephyr_pre0.cmd, line 207.

Does `qemu_cortex_a9` enables `XIP` as a default config?
If not, and if turns out to be too hard to fix, I'd suggest exclusively defining the `SECTION_ALIGN` with `XIP` meaning define `SECTION_ALIGN` with ALIGN(_region_min_align) only if `XIP` is `n` for simplicity, otherwise define an empty `SECTION_ALIGN`.

Any other suggestions?

microbuilder commented:
> If not, and if turns out to be too hard to fix, I'd suggest exclusively defining the `SECTION_ALIGN` with `XIP` meaning define `SECTION_ALIGN` with ALIGN(_region_min_align) only if `XIP` is `n` for simplicity, otherwise define an empty `SECTION_ALIGN`.

This seems reasonable to me as a first step.

ibirnbaum commented:
Sorry that the progress on this has been a tad slow recently.

Changes applied to the commit:
- As suggested by @microbuilder, `SECTION_ALIGN` is now only defined when not building for XIP mode. As this mode of aligning sections is now gone for XIP, each of the two instances where this macro is used is now preceeded by `. = ALIGN (_region_min_align)`. I'll test if this can be made conditional for XIP only, in general, I still have to check if the mapfile generated when building for XIP mode makes any sense. While I do have a Zynq-based board capable of booting from flash, I still have to find out if I can interrupt booting from flash so that I can attach my debugger prior to Zepyhr starting execution.
Still pending:
- fix the z_mapped_start marker

Pushing the update at this stage in order to find out what the CI thinks about this. I checked one self-written test case (which defines a large heap & a large array in bss, so wrong section placement would definitely reflect in output binary file size) against the memory map shown above, and while some sizes have changed a little, the layout remains plausible.

ibirnbaum committed:
arch: aarch32: fix z_mapped_start location for non-XIP operation

Move the z_mapped_start marker for non-XIP builds to the base RAM address.
This has already been the marker's location when building for XIP.

Prior to this change, z_mapped_start was located at the start of the
text section for non-XIP builds. However, at least for the Cortex-A
family of CPUs, the first section located at the RAM base address
isn't the text section, but the single 4k page containing the exception
vectors which are copied to either address 0 or the HIVECS address upon
early boot.

This resulted in this first 4k page at the bottom of RAM to be considered
available for memory mappings via the MMU at run-time, followed by all
the permanently mapped stuff, with available mappable memory only
continuing behind z_mapped_end. A call at run-time requesting memory to
be mapped via the MMU therefore always assigned the single 4k page
containing the vectors data first, before continuing mapping physical
memory behind z_mapped_end. For any map call requesting more than 4k,
this resulted in a contiguous virtual memory address range mapped to
non-contiguous physical memory. This behaviour has already been
documented in #51024.

This behaviour would also cause further problems in case of support
for the Cortex-A's VBAR register, which has been proposed before, but
eventually wasn't merged. Letting VBAR point to the RAM base where
the 4k vectors page is located within the Zephyr image instead of
copying the vectors to either 0 or the HIVECS address means that this
page may under no circumstance be re-assigned for use as regular RAM
mapped via the MMU.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

ibirnbaum committed:
doc: release-notes: document aarch32 cortex_a_r linker command file changes

Document the changes implemented in #60368:
* Placement of the .bss and .noinit sections at the end of the binary
  so that large zero-/uninitialized data structures such as heaps,
  arrays etc. don't have to be padded in the resulting binary.
* Location of the z_mapped_start marker: prevents the assignment
  of the single 4k-page wide .vectors section right at the RAM
  base address as dynamic memory by the MMU at run-time. Instead
  of pointing to the start of the subsequent .text section, the
  z_mapped_start marker now covers all the data contained within
  the binary that ends up in RAM.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

ibirnbaum commented:
The CI put up a bit of a fight last night with misleading 'trailing whitespace' errors, but by now, everything should be good to go.

microbuilder reviewed (approved):
Thanks for taking the time to add release notes as well.
